lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,88 +3641,998 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+
+  void setTangentBuffer(SILBasicBlock *origBB, SILValue originalBuffer,
+                        SILValue adjointBuffer) {
+    assert(originalBuffer->getType().isAddress());
+    auto insertion =
+        bufferMap.try_emplace({origBB, originalBuffer}, adjointBuffer);
+    assert(insertion.second); (void)insertion;
+  }
+
+  SILValue &getTangentBuffer(SILBasicBlock *origBB, SILValue originalBuffer) {
+    auto diffBuilder = getDifferentialBuilder();
+    assert(originalBuffer->getType().isAddress());
+    assert(originalBuffer->getFunction() == original);
+    auto insertion = bufferMap.try_emplace({origBB, originalBuffer},
+                                           SILValue());
+    if (!insertion.second) // not inserted
+      return insertion.first->getSecond();
+
+    // Set insertion point for local allocation builder: before the last local
+    // allocation, or at the start of the adjoint function's entry if no local
+    // allocations exist yet.
+    diffLocalAllocBuilder.setInsertionPoint(
+        getDifferential().getEntryBlock(),
+        getNextDifferentialLocalAllocationInsertionPoint());
+
+    // Allocate local buffer and initialize to zero.
+    auto bufObjectType = getRemappedTangentType(originalBuffer->getType());
+    auto *newBuf = diffLocalAllocBuilder.createAllocStack(
+        originalBuffer.getLoc(), bufObjectType);
+
+    // Temporarily change global builder insertion point and emit zero into the
+    // local buffer.
+    auto insertionPoint = diffLocalAllocBuilder.getInsertionBB();
+    diffBuilder.setInsertionPoint(
+        diffLocalAllocBuilder.getInsertionBB(),
+        diffLocalAllocBuilder.getInsertionPoint());
+    emitZeroIndirect(bufObjectType.getASTType(), newBuf, newBuf->getLoc());
+    diffBuilder.setInsertionPoint(insertionPoint);
+
+    // Create cleanup for local buffer.
+    differentialLocalAllocations.push_back(newBuf);
+    return (insertion.first->getSecond() = newBuf);
+  }
+
+  //--------------------------------------------------------------------------//
+  // Type transformer
+  //--------------------------------------------------------------------------//
+
+  Optional<VectorSpace> getTangentSpace(CanType type) {
+    return type->getAutoDiffAssociatedTangentSpace(
+        LookUpConformanceInModule(getModule().getSwiftModule()));
+  }
+
+  /// Assuming the given type conforms to `Differentiable` after remapping,
+  /// returns the associated tangent space type.
+  SILType getRemappedTangentType(SILType type) {
+    return SILType::getPrimitiveType(
+        getTangentSpace(remapType(type).getASTType())->getCanonicalType(),
+        type.getCategory());
+  }
+
+  //--------------------------------------------------------------------------//
+  // Managed tangent value mapping
+  //--------------------------------------------------------------------------//
+
+  AdjointValue makeZeroTangentValue(SILType type) {
```

Comment:
This does not belong in "tangent value mapping". In `PullbackEmitter` this is part of "adjoint value factory methods".

Reply:
Added:
```swift
  //--------------------------------------------------------------------------//
  // Tangent value factory methods
  //--------------------------------------------------------------------------//

  AdjointValue makeZeroTangentValue(SILType type) { ... }

  AdjointValue makeConcreteTangentValue(SILValue value) { ... }
```